### PR TITLE
[FEATURE] cereal

### PR DESCRIPTION
**Pull request by eseiler**
Adds cereal support (and equality comparison) for SDSL data structures.

* Adds a test header for cereal tests
* Enables tests for containers using cereal (bitcompressed_vector)
* Enables serialisation via cereal for FM-Index
* Removes `load`/`store` function from FM-Index and FM-Index concept

**TODO**
* Would be nice to have a cereal tutorial or dedicated page on the documentation (other PR)

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [submodule "sdsl-lite"]
 	path = submodules/sdsl-lite
-	url = https://github.com/xxsds/sdsl-lite
+	url = https://github.com/eseiler/sdsl-lite
```
Comment (h-2) [inline .gitmodules]: This was not supposed to be committed, right?

File: doc/tutorial/read_mapper/index.md [hunk 2]
```diff
@@ -236,6 +236,6 @@ This is the alignment_file_output construction:
 \snippet doc/tutorial/read_mapper/read_mapper_step4.cpp solution
 Here is the complete program:
 \hint
-\include doc/tutorial/read_mapper/read_mapper_step4.cpp
+\snippet doc/tutorial/read_mapper/read_mapper_step4.cpp complete
```
Comment (h-2) [inline doc/tutorial/read_mapper/index.md]: Why these changes?

Comment (h-2) [reply doc/tutorial/read_mapper/index.md]: ok, got it

File: include/seqan3/search/fm_index/detail/csa_alphabet_strategy.hpp [hunk 3]
```diff
@@ -162,6 +162,30 @@ namespace sdsl
             m_C.load(in);
             read_member(m_sigma, in);
         }
+
+        template <typename archive_t>
+        void CEREAL_SAVE_FUNCTION_NAME(archive_t & ar) const
+        {
+            ar(CEREAL_NVP(m_C));
+            ar(CEREAL_NVP(m_sigma));
```
Comment (h-2) [inline include/seqan3/search/fm_index/detail/csa_alphabet_strategy.hpp]: doesn't NVP mean named-value-pair? Where is the name?

File: include/seqan3/search/fm_index/detail/csa_alphabet_strategy.hpp [hunk 4]
```diff
@@ -162,6 +162,30 @@ namespace sdsl
...
+        {
+            return (m_C == other.m_C) && (m_sigma == other.m_sigma);
+        }
+
+        bool operator!=(plain_byte_alphabet const & other) const noexcept
+        {
+            return !(*this == other);
+        }
         //!\endcond
     };
```
Comment (h-2) [inline include/seqan3/search/fm_index/detail/csa_alphabet_strategy.hpp]: (Why) is this file still necessary?

File: include/seqan3/search/fm_index/fm_index.hpp [hunk 5]
```diff
@@ -388,105 +388,76 @@ class fm_index
...
      * ### Exceptions
      *
      * No-throw guarantee.
      */
-    cursor_type begin() const noexcept
+    bool operator==(fm_index const & rhs) const noexcept
     {
-        return {*this};
+        return (index == rhs.index) && (text_begin == rhs.text_begin) && (text_begin_ss == rhs.text_begin_ss) &&
+               (text_begin_rs == rhs.text_begin_rs);
```
Comment (h-2) [inline include/seqan3/search/fm_index/fm_index.hpp]: What is being compared? Why is it linear?

File: .gitmodules (hunk 1 above)
Comment (eseiler) [reply .gitmodules]: Nope, I just wanted to check if CI runs through with my SDSL PR

File: include/seqan3/search/fm_index/detail/csa_alphabet_strategy.hpp (hunk 3 above)
Comment (eseiler) [reply include/seqan3/search/fm_index/detail/csa_alphabet_strategy.hpp]: the name is the variable name.
The macro ist short hand for, e.g., `cereal::make_nvp("m_C", m_C);`

File: include/seqan3/search/fm_index/detail/csa_alphabet_strategy.hpp (hunk 4 above)
Comment (eseiler) [reply include/seqan3/search/fm_index/detail/csa_alphabet_strategy.hpp]: In general?
The SDSL alphabet strategies map the occurring symbols to ranks. This reduces the memory consumption if not every symbol occurs, but increases the runtime (additional lookup in the map).
This file uses identity (maps to the rank), hence faster.

Though I'd argue this could probably go into the SDSL.

File: include/seqan3/search/fm_index/fm_index.hpp (hunk 5 above)
Comment (eseiler) [reply include/seqan3/search/fm_index/fm_index.hpp]: We're comparing the underlying int_vector representation and the test begin positions.
Since comparing the vectors needs to compare the elements, the complexity is linear relative to its size, no?

File: include/seqan3/search/fm_index/detail/csa_alphabet_strategy.hpp (hunk 3 above)
Comment (h-2) [reply include/seqan3/search/fm_index/detail/csa_alphabet_strategy.hpp]: Yes, but I mean, do you need the names if they are as useless as they currently are? :wink:

File: include/seqan3/search/fm_index/detail/csa_alphabet_strategy.hpp (hunk 4 above)
Comment (h-2) [reply include/seqan3/search/fm_index/detail/csa_alphabet_strategy.hpp]: > Though I'd argue this could probably go into the SDSL.

That's what I meant. Can you open a card for this?

File: include/seqan3/search/fm_index/fm_index.hpp (hunk 5 above)
Comment (h-2) [reply include/seqan3/search/fm_index/fm_index.hpp]: True. But do you need to compare the actual index? Aren't two indexes of the same text identical?

File: test/include/seqan3/test/cereal.hpp [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides cereal functionality for tests.
+ * \author Enrico Seiler <enrico.seiler AT fu-berlin.de>
+ */
+
+#pragma once
+#if SEQAN3_WITH_CEREAL
```
Comment (h-2) [inline test/include/seqan3/test/cereal.hpp]: You need to include seqan3/platform.hpp for this macro to be set

File: test/include/seqan3/test/cereal.hpp [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+namespace test
+{
+
+template <typename in_archive_t, typename out_archive_t, typename TypeParam>
+void do_serialisation(TypeParam const l)
+{
+    // This makes sure the file is also deleted if an exception is thrown in one of the tests below.
+    // Generate unique file name.
+    tmp_filename filename{"cereal_test"};
+    {
```
Comment (h-2) [inline test/include/seqan3/test/cereal.hpp]: newline before

File: test/include/seqan3/test/cereal.hpp [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        in_archive_t iarchive{is};
+        iarchive(in_l);
+        EXPECT_EQ(l, in_l);
+    }
+}
+
+} // namespace seqan3::test
+
+} //namespace seqan3
+#endif // SEQAN3_WITH_CEREAL
```
Comment (h-2) [inline test/include/seqan3/test/cereal.hpp]: can you define these functions as NOPs if cereal isn't defined? Than we can save the checks in every test and just do it here.

File: include/seqan3/search/fm_index/fm_index.hpp (hunk 5 above)
Comment (eseiler) [reply include/seqan3/search/fm_index/fm_index.hpp]: Mhm, then I would have to compare the texts. Still linear?
And they could use different traits I guess. Then the indices would be different for the same texts.

File: test/include/seqan3/test/cereal.hpp (hunk 6 above)
Comment (eseiler) [reply test/include/seqan3/test/cereal.hpp]: 💡

File: test/include/seqan3/test/cereal.hpp (hunk 8 above)
Comment (eseiler) [reply test/include/seqan3/test/cereal.hpp]: Good idea, will do

File: include/seqan3/search/fm_index/detail/csa_alphabet_strategy.hpp (hunk 3 above)
Comment (eseiler) [reply include/seqan3/search/fm_index/detail/csa_alphabet_strategy.hpp]: Expressive uselessness? 🤷‍♂️

File: test/include/seqan3/test/cereal.hpp (hunk 8 above)
Comment (eseiler) [reply test/include/seqan3/test/cereal.hpp]: But that would mean testing every archive in general, but this should be fine, right?

File: include/seqan3/search/fm_index/fm_index.hpp (hunk 5 above)
Comment (h-2) [reply include/seqan3/search/fm_index/fm_index.hpp]: Well, my thought was, that 
  * if the indexes are guaranteed to be the same for the same text
  * and you store a pointer to the text
  * and the pointer value is identical
→ the index is identical and can be compared in `O(1)`. 

You could opt for a less strict definition of identity and compare the contents of the text instead. Or even say that the text is irrelevant and compare the contents of the index, but from my POV that would be unintuitive, because indexes over different alphabets could be equal if they have the same composition of ranks.

File: test/include/seqan3/test/cereal.hpp (hunk 8 above)
Comment (h-2) [reply test/include/seqan3/test/cereal.hpp]: You mean always testing textarchives and binaryarchives? Yes, that would be desirable even, I think.

File: include/seqan3/search/fm_index/fm_index.hpp (hunk 5 above)
Comment (eseiler) [reply include/seqan3/search/fm_index/fm_index.hpp]: Mhm, the problem is that 

1) You could have the same text twice and construct an index each and the indices would not be equal though they are.
2) The text is only a pointer, i.e., when I load an index, the text is a `nullptr` - so I can't really compare the text.

Comment (h-2) [reply include/seqan3/search/fm_index/fm_index.hpp]: > You could have the same text twice and construct an index each and the indices would not be equal though they are.

Yes, if you want this behaviour you need to compare the texts.

> The text is only a pointer, i.e., when I load an index, the text is a nullptr - so I can't really compare the text.

Is the index in a valid state if the pointer is a nullptr?

Comment (eseiler) [reply include/seqan3/search/fm_index/fm_index.hpp]: The index is.
The only thing that doesn't work is to ask the `fm_index_cursor` or such for the `query()` since it needs text access for it.
The search works as expected since it only returns text positions and doesn't access the indexed text.

Comment (eseiler) [reply include/seqan3/search/fm_index/fm_index.hpp]: Btw

>because indexes over different alphabets could be equal if they have the same composition of ranks.

the `fm_index` is templatised over the text type. So we would have to have another `operator==` anyway to compare, e.g., `fm_index<dna4_vector>` and `fm_index<std::string>`?

Comment (h-2) [reply include/seqan3/search/fm_index/fm_index.hpp]: > The index is.

Hm, we need to discuss this at the strategy meeting. I thought certain indexes also do in-text verification. That would not work, right?

> the fm_index is templatised over the text type. So we would have to have another operator== anyway to compare e.g., fm_index<dna4_vector> and fm_index<std::string>?

I was just listing different notions of equality above, I don't thing this should really be possible.

Comment (h-2) [reply include/seqan3/search/fm_index/fm_index.hpp]: You cannot compare dna4_vector with std::string, why should you be able to compare indexes of either with each other?

Comment (eseiler) [reply include/seqan3/search/fm_index/fm_index.hpp]: > Hm, we need to discuss this at the strategy meeting. I thought certain indexes also do in-text verification. That would not work, right?

Well, you still can, but you need to provide the text. The index doesn't own the text and hence it's not serialised with it - that's what we decided on when the fm_index was implemented.

> You cannot compare dna4_vector with std::string, why should you be able to compare indexes of either with each other?

You started the thing with having indices over different alphabets 😀

File: test/include/seqan3/test/cereal.hpp [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+/*!\brief Tests if an object is serialise for all cereal archive types.
+ * \tparam value_t The type to serialise.
+ * \param l The object to serialise.
+ *
+ * If cereal is **not** available, this function is a NOP.
+ * Otherwise it will call do_cerealisation() with cereal's `Binary`, `PortableBinary`, `JSON` and `XML` archives.
+ */
+template <typename value_t>
+void do_serialisation([[maybe_unused]] value_t const l)
```
Comment (h-2) [inline test/include/seqan3/test/cereal.hpp]: this is just testing, but can you do no const and && in any case?

Comment (h-2) [reply test/include/seqan3/test/cereal.hpp]: also above

Comment (eseiler) [reply test/include/seqan3/test/cereal.hpp]: Ok, done

**Comment by h-2** [comment]
fail
